Objects/dictobject.c

```diff
@@ -1709,7 +1734,34 @@ insertdict(PyInterpreterState *interp, PyDictObject *mp,
         assert(mp->ma_keys->dk_kind == DICT_KEYS_GENERAL);
     }
 
-    Py_ssize_t ix = _Py_dict_lookup(mp, key, hash, &old_value);
+    Py_ssize_t ix;
+#ifdef Py_GIL_DISABLED
+    PyDictKeysObject *dk = mp->ma_keys;
+    DictKeysKind kind = dk->dk_kind;
+
+    if (kind == DICT_KEYS_SPLIT) {
```

Comment:
I think there's still a bug where two threads:

1) Check if the key exists using `unicodekeys_lookup_unicode_threadsafe`
2) Both threads then call `insert_split_dict`

Currently, we hold the lock for (1) and (2), but release it in between.

Here is a reproducer I used on the `nogil-integration` branch:
https://gist.github.com/colesbury/fe9d0a9cb8efed46122c54794e79a86b

Reply:
This is also probably a problem with `setdefault` as well.

I think we want to avoid locking the shared keys on insert if we can, so I think that means we need to try the thread safe lookup first, and then if that fails lock the shared keys and re-try the lookup again.

One slight wrinkle is that to get the correct dictionary watcher events delivered we need to deliver those with the shared keys locked.  The things you're supposed to do when getting those events are supposed to be limited, but this feels slightly more dangerous then just having the dictionary locked.  We could potentially deliver the event outside of the lock with the possibility that we deliver an extra event - presumably that might lead to sending multiple adds for the same key as well as sending an add and a modified.